File: backend-go/internal/registry/pypi.go
package registry

import (
	"context"
	"fmt"
	"regexp"
	"strings"
)

var pypiNameRe = regexp.MustCompile(`^[A-Za-z0-9]([A-Za-z0-9._-]*[A-Za-z0-9])?$`)

// PyPI — плагин менеджера pypi. Формат записи: name==version.
type PyPI struct {
	BaseURL string
	HTTP    Doer
}

func (*PyPI) Code() string         { return "pypi" }
func (*PyPI) Title() string        { return "PyPI (Python)" }
func (*PyPI) EntryFormat() string  { return "name==version" }
func (*PyPI) OSVEcosystem() string { return "PyPI" }

// NormalizeName — PEP 503: регистр не важен, разделители эквивалентны.
func (*PyPI) NormalizeName(name string) string {
	return strings.ToLower(pypiSeparators.ReplaceAllString(strings.TrimSpace(name), "-"))
}

var pypiSeparators = regexp.MustCompile(`[-_.]+`)

// NormalizeVersion — полноценная нормализация PEP 440 (1.0 == 1.0.0 == 1.0.0.0)
// в Go потребовала бы отдельного компаратора и здесь сознательно не сделана:
// версия приводится к нижнему регистру без пробелов, а сверка эквивалентных
// записей — задача компаратора версий, который переносится вместе с разбором
// файлов зависимостей. Для конвейера этого достаточно: версия приходит из
// заявки и уходит в реестр как есть.
func (*PyPI) NormalizeVersion(version string) string {
	return strings.ToLower(strings.TrimSpace(version))
}

func (*PyPI) DisplayName(name string) string { return strings.TrimSpace(name) }

func (p *PyPI) SplitEntry(entry string) (string, string, error) {
	text := strings.TrimSpace(entry)
	name, version, found := strings.Cut(text, "==")
	if !found {
		return "", "", invalidFormat(p.EntryFormat(),
			"«%s» не соответствует формату pypi. Ожидается: name==version "+
				"(например, requests==2.31.0)", text)
	}
	return strings.TrimSpace(name), strings.TrimSpace(version), nil
}

func (p *PyPI) ValidateName(name string) error {
	if len(name) > 512 {
		return invalidFormat(p.EntryFormat(), "Имя пакета длиннее 512 символов")
	}
	if !pypiNameRe.MatchString(name) {
		return invalidFormat(p.EntryFormat(),
			"Недопустимое имя пакета pypi: «%s» (PEP 503: буквы, цифры, «-», «_», «.»)", name)
	}
	return nil
}

var pypiVersionRe = regexp.MustCompile(`^[0-9]+(\.[0-9]+)*((a|b|rc|alpha|beta|post|dev)\.?[0-9]*)*(\.post[0-9]+)?(\.dev[0-9]+)?(\+[A-Za-z0-9.]+)?$`)

func (p *PyPI) ValidateVersion(version string) error {
	if len(version) > 128 {
		return invalidFormat(p.EntryFormat(), "Версия длиннее 128 символов")
	}
	if !pypiVersionRe.MatchString(strings.ToLower(strings.TrimSpace(version))) {
		return invalidFormat(p.EntryFormat(),
			"Версия «%s» не соответствует PEP 440 (например, 2.31.0, 1.0.0rc1)", version)
	}
	return nil
}

// pypiResponse — та часть ответа /pypi/{name}/{version}/json, что нам нужна.
type pypiResponse struct {
	Info struct {
		License           string   `json:"license"`
		LicenseExpression string   `json:"license_expression"`
		Classifiers       []string `json:"classifiers"`
		Yanked            bool     `json:"yanked"`
	} `json:"info"`
	URLs []pypiDist `json:"urls"`
}

type pypiDist struct {
	PackageType       string            `json:"packagetype"`
	PythonVersion     string            `json:"python_version"`
	URL               string            `json:"url"`
	Filename          string            `json:"filename"`
	Size              int64             `json:"size"`
	Yanked            bool              `json:"yanked"`
	UploadTime        string            `json:"upload_time"`
	UploadTimeISO8601 string            `json:"upload_time_iso_8601"`
	Digests           map[string]string `json:"digests"`
}

func (p *PyPI) FetchMetadata(ctx context.Context, ref Ref) (Metadata, error) {
	url := fmt.Sprintf("%s/pypi/%s/%s/json", p.BaseURL, ref.Name, ref.RawVersion)
	var payload pypiResponse
	if err := getJSON(ctx, p.HTTP, url, "application/json", &payload); err != nil {
		if err == ErrNotFound {
			return Metadata{}, fmt.Errorf("%w: пакет %s==%s отсутствует в реестре pypi",
				ErrNotFound, ref.DisplayName, ref.RawVersion)
		}
		return Metadata{}, err
	}

	chosen := choosePypiDist(payload.URLs)
	meta := Metadata{Name: ref.Name, Version: ref.RawVersion}

	if chosen != nil {
		meta.PublishedAt = parseTime(chosen.UploadTimeISO8601)
		if meta.PublishedAt == nil {
			meta.PublishedAt = parseTime(chosen.UploadTime)
		}
		meta.ArtifactURL = chosen.URL
		meta.ArtifactFilename = chosen.Filename
		meta.SizeBytes = chosen.Size
		meta.Yanked = chosen.Yanked
		if sha := chosen.Digests["sha256"]; sha != "" {
			meta.Checksum, meta.ChecksumAlgo = sha, "sha256"
		}
	}
	// Дата публикации любой дистрибуции годится: карантин считается по версии,
	// а не по конкретному файлу.
	if meta.PublishedAt == nil {
		for i := range payload.URLs {
			if t := parseTime(payload.URLs[i].UploadTimeISO8601); t != nil {
				meta.PublishedAt = t
				break
			}
			if t := parseTime(payload.URLs[i].UploadTime); t != nil {
				meta.PublishedAt = t
				break
			}
		}
	}
	meta.Yanked = meta.Yanked || payload.Info.Yanked

	meta.LicenseRaw = strings.TrimSpace(payload.Info.License)
	meta.LicenseSPDX = NormalizeSPDX(payload.Info.LicenseExpression)
	if meta.LicenseSPDX == "" {
		meta.LicenseSPDX = NormalizeSPDX(meta.LicenseRaw)
	}
	if meta.LicenseSPDX == "" {
		meta.LicenseSPDX = SPDXFromClassifiers(payload.Info.Classifiers)
	}
	return meta, nil
}

// choosePypiDist предпочитает wheel: он же обычно и публикуется во внутренний
// репозиторий.
func choosePypiDist(urls []pypiDist) *pypiDist {
	var wheels, sdists []*pypiDist
	for i := range urls {
		switch urls[i].PackageType {
		case "bdist_wheel":
			wheels = append(wheels, &urls[i])
		case "sdist":
			sdists = append(sdists, &urls[i])
		}
	}
	for _, w := range wheels {
		if strings.Contains(w.PythonVersion, "py3") {
			return w
		}
	}
	if len(wheels) > 0 {
		return wheels[0]
	}
	if len(sdists) > 0 {
		return sdists[0]
	}
	if len(urls) > 0 {
		return &urls[0]
	}
	return nil
}

func (p *PyPI) InstallCommand(ref Ref, baseURL, repo string) string {
	index := fmt.Sprintf("%s/repository/%s/simple", strings.TrimRight(baseURL, "/"), repo)
	return fmt.Sprintf("pip install -i %s %s==%s", index, ref.DisplayName, ref.RawVersion)
}

func (p *PyPI) ArtifactPath(ref Ref, filename string) string {
	return fmt.Sprintf("%s/%s/%s", ref.Name, ref.RawVersion, filename)
}

File: backend/app/managers/pypi.py
"""Плагин менеджера pypi. Формат записи: `name==version`."""

from __future__ import annotations

import re
from datetime import datetime
from typing import Any

from app.core.config import get_settings
from app.core.errors import InvalidPackageFormat, UpstreamError
from app.core.http import client, request_with_retries
from app.managers.base import PackageManagerPlugin, PackageRef, ParsedEntry, RegistryMetadata
from app.managers.parsers.python_files import (
    parse_poetry_lock,
    parse_pyproject_toml,
    parse_requirements_txt,
)
from app.managers.utils import parse_iso8601, raw_deps_to_entries
from app.services.spdx import normalize_spdx, spdx_from_classifiers

_NAME_RE = re.compile(r"^[A-Za-z0-9]([A-Za-z0-9._-]*[A-Za-z0-9])?$")


class PypiPlugin(PackageManagerPlugin):
    code = "pypi"
    title = "PyPI (Python)"
    entry_format = "name==version"
    dependency_files = ("requirements*.txt", "poetry.lock", "pyproject.toml")
    # Формат не различает прямые и транзитивные — см. golang.py.
    indeterminate_kind_files = ("poetry.lock",)

    # ------------------------------------------------------------------ нормализация
    def normalize_name(self, name: str) -> str:
        # PEP 503: регистр не важен, разделители эквивалентны
        return re.sub(r"[-_.]+", "-", name.strip()).lower()

    def normalize_version(self, version: str) -> str:
        from packaging.version import InvalidVersion, Version

        version = version.strip()
        try:
            return str(Version(version))
        except InvalidVersion:
            return version

    def split_entry(self, entry: str) -> tuple[str, str]:
        text = entry.strip()
        if "==" not in text:
            raise InvalidPackageFormat(
                f"«{text}» не соответствует формату pypi. Ожидается: name==version "
                "(например, requests==2.31.0)",
                expected_format=self.entry_format,
            )
        name, _, version = text.partition("==")
        return name.strip(), version.strip()

    def validate_name(self, name: str) -> None:
        super().validate_name(name)
        if not _NAME_RE.match(name):
            raise InvalidPackageFormat(
                f"Недопустимое имя пакета pypi: «{name}» (PEP 503: буквы, цифры, «-», «_», «.»)"
            )

    def validate_version(self, version: str) -> None:
        from packaging.version import InvalidVersion, Version

        super().validate_version(version)
        try:
            Version(version)
        except InvalidVersion as exc:
            raise InvalidPackageFormat(
                f"Версия «{version}» не соответствует PEP 440 (например, 2.31.0, 1.0.0rc1)"
            ) from exc

    # ------------------------------------------------------------------ файлы
    def parse_dependency_file(self, filename: str, content: bytes) -> list[ParsedEntry]:
        base = filename.rsplit("/", 1)[-1]
        if base == "poetry.lock":
            raw = parse_poetry_lock(content)
        elif base == "pyproject.toml":
            raw = parse_pyproject_toml(content)
        elif base.startswith("requirements") and base.endswith(".txt"):
            raw = parse_requirements_txt(content)
        else:
            raise InvalidPackageFormat(
                f"Файл «{base}» не поддерживается для pypi. Поддерживаются: "
                "requirements.txt, poetry.lock, pyproject.toml"
            )
        return raw_deps_to_entries(self, raw)

    # ------------------------------------------------------------------ реестр
    def fetch_metadata(self, ref: PackageRef) -> RegistryMetadata:
        base = get_settings().registry_pypi_url.rstrip("/")
        url = f"{base}/pypi/{ref.name}/{ref.raw_version}/json"
        with client() as http:
            resp = request_with_retries("pypi", lambda: http.get(url, headers={"Accept": "application/json"}))
        if resp.status_code == 404:
            raise UpstreamError(
                f"Пакет {ref.display_name}=={ref.raw_version} не найден в реестре pypi",
                registry="pypi",
            )
        if resp.status_code >= 400:
            raise UpstreamError(f"Реестр pypi ответил {resp.status_code}", registry="pypi")
        return self._parse_metadata(ref, resp.json())

    def _parse_metadata(self, ref: PackageRef, payload: dict[str, Any]) -> RegistryMetadata:
        info = payload.get("info") or {}
        urls = payload.get("urls") or []
        chosen = self._choose_dist(urls)

        published: datetime | None = None
        if chosen:
            published = parse_iso8601(chosen.get("upload_time_iso_8601") or chosen.get("upload_time"))
        if published is None:
            for entry in urls:
                published = parse_iso8601(entry.get("upload_time_iso_8601") or entry.get("upload_time"))
                if published:
                    break

        license_raw = (info.get("license") or "").strip() or None
        spdx = normalize_spdx(info.get("license_expression") or "") or normalize_spdx(license_raw or "")
        if not spdx:
            spdx = spdx_from_classifiers(info.get("classifiers") or [])

        digests = (chosen or {}).get("digests") or {}
        return RegistryMetadata(
            name=ref.name,
            version=ref.raw_version,
            published_at=published,
            license_spdx=spdx,
            license_raw=license_raw,
            artifact_url=(chosen or {}).get("url"),
            artifact_filename=(chosen or {}).get("filename"),
            checksum=digests.get("sha256"),
            checksum_algo="sha256" if digests.get("sha256") else None,
            size_bytes=(chosen or {}).get("size"),
            yanked=bool((chosen or {}).get("yanked") or info.get("yanked")),
            raw={"info": {k: info.get(k) for k in ("summary", "home_page", "project_urls")}},
        )

    @staticmethod
    def _choose_dist(urls: list[dict[str, Any]]) -> dict[str, Any] | None:
        """Предпочитаем wheel: он же обычно и публикуется во внутренний репозиторий."""
        wheels = [u for u in urls if u.get("packagetype") == "bdist_wheel"]
        if wheels:
            py3 = [u for u in wheels if "py3" in (u.get("python_version") or "")]
            return (py3 or wheels)[0]
        sdists = [u for u in urls if u.get("packagetype") == "sdist"]
        return (sdists or urls or [None])[0]

    # ------------------------------------------------------------------ публикация
    def publish_target(self, ref: PackageRef, filename: str) -> dict[str, str]:
        return {"format": "pypi", "filename": filename, "asset_field": "pypi.asset"}

    def install_command(self, ref: PackageRef, base_url: str, repo: str) -> str:
        index = f"{base_url.rstrip('/')}/repository/{repo}/simple"
        return f"pip install -i {index} {ref.display_name}=={ref.raw_version}"
